### ISSUE TITLE
Alien Swarm

### ISSUE THREAD

**Issue by UniverseTech** [ValveSoftware/steam-for-linux#2254]
Hello, there is such game Alien Swarm, please, release the server for Linux! (now only under win)

**Comment by aaronfranke** [comment]
October 2015, Valve's promoting Linux so much, 1550 total Linux games, why isn't Valve's own Alien Swarm not one of them?

**Comment by testman42** [comment]
@gdrewb-valve Any updates regarding this feature request?

**Comment by gdrewb-valve** [comment]
I haven't heard anything either way, so no there's no news.

**Comment by samzanemesis** [comment]
@testman42 Alien Swarm seems to be EOL'd, even though it's technically the most up-to-date public branch compared to their latest games it's missing a lot of critical security patches

**Comment by testman42** [comment]
Could they release it to community for maintenance then?
Alien Swarm was never among Valve's well known IPs (I'd say its somewhere around Ricochet), and since Valve said making games isn't their main focus any more (correct me if I'm wrong), allowing community to maintain and update game would be best case scenario. They are already doing something similar with TF2, where community has very big input in game's development.

**Comment by aaronfranke** [comment]
Agreed, that sounds like a great idea. Turn it into a source mod or something.

**Comment by Cheeseness** [comment]
That's effectively what the Alien Swarm is, and the [Alien Swarm SDK](https://developer.valvesoftware.com/wiki/Authoring_Tools/SDK_%28Alien_Swarm%29) provides source for the game specific DLLs. I've been hoping for the past couple of years to find time to poke around and see if I could port that to a Linux supporting Source game, but it's low on my priority list.

**Comment by ntvsx193** [comment]
No progress for migrate on Linux?(

**Comment by Yepoleb** [comment]
I recently looked into doing a community port, but it doesn't look like that's possible. We don't just need a game DLL that runs on Linux, but also an engine. Our current options are:

1. Source SDK 2013 Base. Lacks all the features that were implemented after HL2. We'd have to backport 6 years of engine development and patch the binary. Pretty sure that's not going to happen.

2. Hack together an SDK for one of the more "recent" Valve releases. I found this [reddit thread](https://www.reddit.com/r/SourceEngine/comments/36rl3s/updated_source_version_changelog/) speculating how the Source branches developed. Dota, CS:GO are out because they're Source 2 and something in between 1 and 2. Anything based on Orange Box is out as well, like explained above. All that's left now is Portal 2 and Left 4 Dead (2). I have some doubts they're going to be able to run Alien Swarm, but maybe it's worth further investigation. Still, we'd need to reverse engineer our own SDK and that's a ton of work.

3. Backport everything to the leaked Source 2007 code. Basically 1., but binary patching gets replaced with legal risk.

4. Ask for a custom license. No idea how our chances would be. They've been quite helpful towards mod developers who want to release on Steam, but also really secretive about their terms.

I hope this helps everyone who wants to go down the same rabbit hole. The most promising solution at this point would be that someone at Valve regains interest after the release of Reactive Drop and decides to do the port.

**Comment by Cheeseness** [comment]
> We don't just need a game DLL that runs on Linux, but also an engine.

I wasn't meaning to imply that getting a DLL built was the limit of the work, just that the game specific code is present within the Alien Swarm SDK.

L4D2 would be the most likely candidate for porting AS to - Alien Swarm shared a bunch of L4D features (like the AI Director, etc.) and is probably closest to that branch.

I can't imagine it'd require much reverse engineering based on the superficial investigation I'd done, but it'd still be a huge amount of work for anybody without experience working with Source/migrating between Source versions.

**Comment by Yepoleb** [comment]
The SourceMod people have already done quite a bit of work on the SDK, but there's still a lot missing to write an actual mod for L4D2. The thing that worries me the most are potentially missing engine features. They'd probably be really hard to patch in. But if you want to give it a try, I'm in.

**Comment by samzanemesis** [comment]
Left 4 Dead 2 deviates a lot from the branch used in Alien Swarm, the BSP format isn’t even compatible nor are some of the models, the closest match is Portal 2, good luck earning a license for that, plus a license to port ASW to that branch though


The Reactive Drop team doesn’t have the engine source code








De: Gabriel Huber
Enviado: ‎segunda-feira‎, ‎1‎ de ‎maio‎ de ‎2017 ‎15‎:‎47
Para: ValveSoftware/steam-for-linux
Cc: Sam Pavlovic, Comment





The SourceMod people have already done quite a bit of work on the SDK, but there's still a lot missing to write an actual mod for L4D2. The thing that worries me the most are potentially missing engine features. They'd probably be really hard to patch in. But if you want to give it a try, I'm in.

**Comment by aaronfranke** [comment]
An engine upgrade to Source 2013 wouldn't hurt, well, except for developer's available time that is. How much of the code for Alien Swam do you think could be modularized and be made portable?

**Comment by csallen1204** [comment]
I've been taking a stab at shoehorning the Swarm/Reactive Drop SDK into Source 2013 SDK.

The original Swarm SDK makes it evident that it was written primarily for Windows and a lot of things that Visual Studio would let you get away with become issues with Mac/Linux compilers. There are also some files that are making calls to Windows libraries. In some cases like itoa or _countof, it's easy to supersede them with more cross-platform-friendly replacements, but there are other cases like inter-process-semaphores where it gets real ugly. So far I've been lucky that the ones without quick workarounds have been able to get stubbed out on other platforms via ```#ifdef _WIN32``` because they aren't essential to making the game run.

After that, it's a matter of:
- Restoring Alien Swarm-specific customizations to the newer source 2013 sdk files
- Updating older SDK version functions that are missing in the 2013 sdk.
- Updating Swarm files to incorporate 2013 sdk changes

As to how it will run, once I have the libraries built I will let you know.

**Comment by testman42** [comment]
Thank you man, we appreciate it a lot :)

**Comment by csallen1204** [comment]
After going over the code bases for Alien Swarm and the Source 2013 SDK, I've come to the conclusion that, without any assistance from Valve, it's not possible to create a Linux/Mac port.

The Source 2013 SDK, the only SDK that has Linux/Mac support, is based on just Half-Life 2. Other Source games branched off from it long ago and have other components that aren't provided in the 2013 SDK, such as VScript, Response Rules, Logging Channels, etc. Existing components in the 2013 SDK are missing functions that were part of the Alien Swarm SDK. I was able to find them in static libraries from other Source-engine-based games but not all. If it was the case that the core c++ files for these components were available, it would be possible to build the libraries for Linux and Mac, but these code bases are closed off to the public. The Alien Swarm SDK has lib and dll files that cover these but they are only for Windows.

**Comment by aaronfranke** [comment]
What about porting to another version of the Source SDK?

**Comment by samzanemesis** [comment]
The only other versions of the SDK are previous iterations of Orangebox
SDK, not viable either.

I really wish Valve would release an SDK using L4D2 or Portal 2 as base if
they won't release Source 2 for developer, doesn't need game code from
those games, just a stable codebase from a template that's closer to what
they have internally, ASW SDK doesn't even run in certain modern systems
due to lack of updates even if it's the latest public version of the engine.

1. des. 2017 01.54 skrev "Aaron Franke" <notifications@github.com>:

**Comment by csallen1204** [comment]
To elaborate a little, here were some of the things I see missing in the Alien Swarm/Reactive Drop SDK that are crucial to building the libs for Alien Swarm to work in Linux:

-The devtools files for Linux/Mac support
-The vpc files to create the make ones for client, server, and missionchooser
-The Linux library files usually located in lib/common and lib/public

Looking at the platform.h file in public/tier0, I see that this version of the SDK has support for non-Windows operating systems:

```
#ifdef COMPILER_GCC
#define GNUC 1
#endif
```

```
#ifdef PLATFORM_POSIX
```

**Comment by samzanemesis** [comment]
These preprocessor defines are a leftover for Linux servers, something that
was never delivered for ASW either, rip

**Comment by antdude** [comment]
What's the status now?

**Comment by csallen1204** [comment]
@antdude Alien Swarm works great with Proton. You should give that a shot.

**Comment by ghost** [comment]
I've made a heavily modified fork of the Alien Swarm SDK with linux64 and linux32 support: https://github.com/SwarmPorts/AlienSwarmGO.

All you need to do is run the project with CS:GO binaries, everything needed is listed, should be straightforward.
